### ISSUE TITLE
index not bound checked.

### ISSUE THREAD

**Issue by JeremyWildsmith** [CCareaga/heap_allocator#8]
I am getting a crash using this for a personal project, and have determined that it is due an error in heap_alloc.

````
    while (found == NULL) {
        if(index >= BIN_COUNT)
            return NULL;

        temp = heap->bins[++index];
        found = get_best_fit(temp, size);
    }
````

Notice the index increment, and access, without first checking if it is below BIN_COUNT.

**Comment by CCareaga** [comment]
Ahh I must not have checked what happens when the heap is full...

So you are saying the line `index >= BIN_COUNT` should be `index >= BIN_COUNT - 1` ?

**Comment by JeremyWildsmith** [comment]
Whoops, sorry looks like I pasted in my hacked in fix instead of the original code by mistake in my original message. The original code (with no index bound check) is:
````
    while (found == NULL) {
        temp = heap->bins[++index];
        found = get_best_fit(temp, size);
}
````

The hacked in fix I ended up going with was actually:
````

    while (found == NULL) {
        if(index + 1 >= BIN_COUNT)
            return NULL;

        temp = heap->bins[++index];
        found = get_best_fit(temp, size);
    }
````

That seems to have worked for me. Hmm, but honestly I haven't tested it thoroughly enough to really make any guarantee on it. But with that in there, it should be returning null if there is no memory left.

**Comment by CCareaga** [comment]
Oh that looks like it should work, good catch!

You could go ahead and create a pull request, or I could go ahead and fix it.

**Comment by JeremyWildsmith** [comment]
Cool, when I get back in the office on Monday I'll look at making a pr. Thanks!

**Comment by CCareaga** [comment]
resolved by #9
